index.js

```diff
@@ -1,61 +1,112 @@
-function highlightQuery (query, errors) {
-  var locations = errors.map(function (e) { return e.locations })
-    .reduce(function (a, b) {
-      return a.concat(b)
-    }, [])
+/* global fetch, Headers */
+require('isomorphic-fetch')
 
-  var queryHighlight = ''
+function Client (options) {
+  if (!options.url) throw new Error('Missing url parameter')
 
-  query.split('\n').forEach(function (row, index) {
-    var line = index + 1
-    var lineErrors = locations.filter(function (loc) { return loc.line === line })
+  this.options = options
+  this.url = options.url
+  // A stack of registered listeners
+  this.listeners = []
+}
+
+// to reduce file size
+var proto = Client.prototype
+
+/**
+ * Send a query and get a Promise
+ * @param   {String}   query
+ * @param   {Object}   variables
+ * @param   {Function} beforeRequest hook
+ * @returns {Promise}
+ */
+proto.query = function (query, variables, beforeRequest) {
+  var headers = new Headers()
+  headers.set('content-type', 'application/json')
+
+  var req = this.options.request || {}
+  req.method || (req.method = 'POST')
+  req.body || (req.body = JSON.stringify({
+    query: query,
+    variables: variables
+  }))
+  req.headers || (req.headers = headers)
+
+  if (beforeRequest) {
+    var result = beforeRequest(req)
 
-    queryHighlight += row + '\n'
+    // The `beforeRequest` hook may redefine response when returning something
+    if (typeof result !== 'undefined') {
+      return Promise.resolve(result)
+    }
+  }
 
-    if (lineErrors.length) {
-      var errorHighlight = []
+  return this.fetch(req)
+}
 
-      lineErrors.forEach(function (line) {
-        for (var i = 0; i < 8; i++) {
-          errorHighlight[line.column + i] = '~'
-        }
-      })
+/**
+ * For making requests
+ * @param   {Object} req
+ * @returns Promise
+ */
+proto.fetch = function (req) {
+  var self = this
 
-      for (var i = 0; i < errorHighlight.length; i++) {
-        queryHighlight += errorHighlight[i] || ' '
-      }
-      queryHighlight += '\n'
+  var results = self.trigger('request', [req])
```

Comment:
I know that in jQuery they pass an array of arguments instead of passing args as args. But maybe you are right, let's do this the Node way. I fixed this